test/unittest/blas1/blas1_rotmg_test.cpp

```diff
@@ -0,0 +1,256 @@
+
+  /* Validate param */
+  constexpr scalar_t unit_matrix = -2;
+  constexpr scalar_t rescaled_matrix = -1;
+  constexpr scalar_t sltc_matrix = 0;
+  constexpr scalar_t clts_matrix = 1;
+  constexpr scalar_t error_matrix = 2;
+
+  scalar_t flag_sycl = sycl_out.param[0];
+  scalar_t h11_sycl = sycl_out.param[1];
+  scalar_t h12_sycl = sycl_out.param[3];
+  scalar_t h21_sycl = sycl_out.param[2];
+  scalar_t h22_sycl = sycl_out.param[4];
+
+  scalar_t flag_ref = sycl_out.param[0];
+  scalar_t h11_ref = sycl_out.param[1];
+  scalar_t h12_ref = sycl_out.param[3];
+  scalar_t h21_ref = sycl_out.param[2];
+  scalar_t h22_ref = sycl_out.param[4];
+
+  if (flag_sycl != error_matrix && flag_ref != unit_matrix) {
+    if (flag_ref == sltc_matrix) {
+      ASSERT_TRUE(h12_ref == h12_sycl);
+      ASSERT_TRUE(h21_ref == h21_sycl);
+    } else if (flag_ref == clts_matrix) {
+      ASSERT_TRUE(h11_ref == h11_sycl);
+      ASSERT_TRUE(h22_ref == h22_sycl);
+    } else {
+      ASSERT_TRUE(h11_ref == h11_sycl);
+      ASSERT_TRUE(h12_ref == h12_sycl);
+      ASSERT_TRUE(h21_ref == h21_sycl);
+      ASSERT_TRUE(h22_ref == h22_sycl);
+    }
+  }
+};
+
+/*
+ * Rotm can be used to validate that Rotmg outputs will set y to 0. The right
+ * side of the following formula is calculated by rotm.
+ *
+ * x1_output * sqrt(d1_output) = [ h11 h12 ] * [ x1_input]
+ * 0.0       * sqrt(d2_output)   [h21 h22 ]    [ y1_input]
+ */
+template <typename scalar_t>
+void RotmgTest<scalar_t>::validate_with_rotm() {
+  if (sycl_out.param[0] == 2 || sycl_out.d2 < 0) {
+    return;
+  }
+
+  index_t size = 1;
+  index_t incX = 1;
+  index_t incY = 1;
+
+  std::vector<data_t> x_cpu_v{input.x1};
+  std::vector<data_t> y_cpu_v{input.y1};
+
+  reference_blas::rotm(size, x_cpu_v.data(), incX, y_cpu_v.data(), incY,
+                       sycl_out.param.data());
+
+  x_cpu_v[0] = x_cpu_v[0] * static_cast<scalar_t>(sqrt(sycl_out.d1));
+  y_cpu_v[0] = y_cpu_v[0] * static_cast<scalar_t>(sqrt(sycl_out.d2));
+
+  bool y1_becomes_zero = utils::almost_equal<data_t, scalar_t>(y_cpu_v[0], 0);
+  ASSERT_TRUE(y1_becomes_zero);
+}
+
+template <typename scalar_t>
+void run_test(const combination_t<scalar_t> combi) {
+  scalar_t d1_input;
+  scalar_t d2_input;
+  scalar_t x1_input;
+  scalar_t y1_input;
+
+  std::tie(d1_input, d2_input, x1_input, y1_input) = combi;
+
+  RotmgTest<scalar_t> test{d1_input, d2_input, x1_input, y1_input};
+  test.run_sycl_blas_rotmg();
+
+  /* Do not test with things that might overflow or underflow. Results will not
+   * make sense if that happens */
+  if (!test.isOverflowTest()) {
+    test.validate_with_reference();
+    test.validate_with_rotm();
+  }
+}
+
+const auto combi = ::testing::Combine(
+    ::testing::Values(0, 15.5, -2.2, max_float, min_float),  // d1
+    ::testing::Values(0, 3.0, -2.2, max_float, min_float),   // d2
+    ::testing::Values(0, 12.1, -7.3, max_float, min_float),  // x1
+    ::testing::Values(0, 0.5, -4.3, max_float, min_float)    // y1
+);
```

Comment:
There are so many possible code paths that I think you need many more combinations are needed. Ideally we'd need a way to check we have good code coverage.

Reply:
It's quite hard to even think of tests that can trigger every code path. Many of them rely on overflow or underflows  to be triggered.

In addition, I think the current unit test framework is not adequate for rotmg testing. I need to define tests with very specific inputs and using different values for half, float and double (i.e. a templated combi). Instead of combinations of all of them which increases the numbers of tests by a lot.

So, while I can definitely work on it if we agree it makes sense. It will require some effort to update the framework.

Reply:
Nevermind, found a clean way to do this without modifying the test framework. I will add more tests for specific code paths :)

Reply:
Updated the tests to test as many code paths as possible